rust_src/src/casetab.rs

```diff
@@ -3,100 +3,138 @@
 use remacs_macros::lisp_fn;
 
 use crate::{
+    alloc::make_char_table,
     buffers::current_buffer,
     buffers::LispBufferRef,
     chartable::LispCharTableRef,
     lisp::LispObject,
     lists::put,
-    objects::eq,
     remacs_sys::EmacsInt,
     remacs_sys::{
-        map_char_table, set_char_table_extras, set_char_table_purpose, staticpro, Fcopy_sequence,
-        Fmake_char_table, Fset_char_table_range, CHAR_TABLE_SET,
+        map_char_table, staticpro, Fcopy_sequence, Fset_char_table_range, CHAR_TABLE_SET,
     },
     remacs_sys::{Qcase_table, Qcase_table_p, Qchar_table_extra_slots, Qnil},
+    symbols::LispSymbolRef,
     threads::ThreadState,
 };
 
+#[derive(Copy, Clone)]
 pub struct LispCaseTable(LispCharTableRef);
 
 impl LispCaseTable {
+    pub fn convert_char_table(mut table: LispCharTableRef) -> Self {
+        table.purpose = Qcase_table;
+        Self(table)
+    }
+
+    pub fn from_char_table(table: LispCharTableRef) -> Option<Self> {
+        if table.purpose.eq(Qcase_table) {
+            Some(Self(table))
+        } else {
+            None
+        }
+    }
+
+    pub fn is_proper(&self) -> bool {
+        let (up, canon, eqv) = self.extras();
+
+        (up.is_nil() || up.is_char_table())
+            && ((canon.is_nil() && eqv.is_nil())
+                || (canon.is_char_table() && (eqv.is_nil() || eqv.is_char_table())))
+    }
+
     pub fn extras(&self) -> (LispObject, LispObject, LispObject) {
         let extras = unsafe { self.0.extras.as_slice(3) };
         (extras[0], extras[1], extras[2])
     }
 
-    pub fn from_char_table(table: LispCharTableRef) -> Self {
-        Self(table)
+    pub fn set_extras(&mut self, idx: usize, value: LispObject) {
```

Comment:
Renaming this method to `set_extra` would make its purpose more clear. Extra being plural implies all the extras are being set. While on the subject, would it make sense to have a method taking a tuple that sets all of the extras?